Create JS API for Search to set as default

petemill opened the pull request:
Resolves https://github.com/brave/brave-browser/issues/16270

Please see linked security and privacy review(s).


## Submitter Checklist:

- [x] I confirm that ~no security/privacy review [is needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or that~ I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

petemill commented on components/brave_search/common/brave_search_default.mojom:
```diff
@@ -0,0 +1,8 @@
+import "mojo/public/mojom/base/values.mojom";
+
+interface BraveSearchDefault {
+  IsDefaultSearchProvider() => (bool is_default_search_provider);
+  SetIsDefaultSearchProvider();
```
TODO: these functions should act on the _current site_, so should accept the domain. That's so we can maintain functionality on dev / staging / production / local.

bridiver replied:
why wouldn't you get the domain from the render frame?

petemill replied:
Yeah we would. What I mean here is we would pass the domain to the browser host so that the host acts on the current domain - search.brave.com dev-search.brave.com, etc.

bridiver commented on components/brave_search/renderer/brave_search_renderer_utils.cc:
```diff
@@ -0,0 +1,32 @@
+
+namespace brave_search {
+
+bool IsAllowedHost(const GURL& url) {
```
we should also verify these are https so they can't be spoofed

bridiver commented on components/brave_search/renderer/brave_search_default_js_handler.cc:
```diff
@@ -0,0 +1,157 @@
+    // TODO(petemill): Consider showing a confirmation dialog to user.
+    brave_search_default_->SetIsDefaultSearchProvider();
+  } else {
+    blink::WebString message = "setIsDefaultSearchProvider: "
```
seems like this should return an error to the page?

petemill replied:
Well, this is to block malicious or incorrect use of the script i.e. not putting it in a click handler. So logging an error in devtools is ok because the devs will know at dev-time that they haven't put it in the right place. Seems like this is how it's done in other places in chromium 
/ blink.

bridiver commented on components/brave_search/renderer/brave_search_default_js_handler.cc:
```diff
@@ -0,0 +1,157 @@
+
+void BraveSearchDefaultJSHandler::AddJavaScriptObjectToFrame(
+    v8::Local<v8::Context> context) {
+  v8::Isolate* isolate = blink::MainThreadIsolate();
```
I think this is only relevant for the main frame?

bridiver replied:
this class gets created for every frame

bridiver commented on browser/brave_content_browser_client.cc:
```diff
@@ -213,11 +215,22 @@ void BindBraveSearchHost(
+void BindBraveSearchDefaultHost(
+    content::RenderFrameHost* const frame_host,
+    mojo::PendingReceiver<brave_search::mojom::BraveSearchDefault> receiver) {
+  // auto* context = frame_host->GetBrowserContext();
```
this most likely does not apply in private or tor

petemill commented on components/weekly_storage/BUILD.gn:
```diff
@@ -1,5 +1,7 @@
 source_set("weekly_storage") {
   sources = [
+    "daily_storage.cc",
+    "daily_storage.h",
```
I guess this component could be renamed `periodic_storage`

iefremov commented:
cc @simonhong as Simon recently fixed some problems in search engine providers

petemill commented:
`gn_check` now fixed

simonhong commented on components/brave_search/browser/brave_search_default_host.cc:
```diff
@@ -0,0 +1,141 @@
+            << "Site has not created search engine: " << host_;
+    return false;
+  }
+  const auto* default_provider =
```
Recently, I pushed crash fix (https://github.com/brave/brave-core/pull/8930) to handle `nullptr` return value from `TemplateURLService::GetDefaultSearchProvider()`.
I'm not 100% sure that crash came from this `nullptr`. but I think it would be good to handle it here also.

petemill replied:
fixed

simonhong commented on components/brave_search/browser/brave_search_default_host.h:
```diff
@@ -0,0 +1,51 @@
+  ~BraveSearchDefaultHost() override;
+
+  // brave_search::mojom::BraveSearchDefault:
+  void GetCanSetDefaultSearchProvider(
```
How this works if user installed search provider extension such as(https://chrome.google.com/webstore/detail/ecosia-the-search-engine/eedlgdlajadkbbjoobobefphmfkcchfk)?
IMO, if user installed this kind of extension, maybe we should not change default search provider?

petemill replied:
also group policy via `is_default_search_managed()`

petemill replied:
fixed

iefremov commented:
@petemill do we have a spec or a ticket for this work?

iefremov commented on components/brave_search/renderer/brave_search_renderer_utils.cc:
```diff
@@ -0,0 +1,36 @@
+
+namespace {
+
+static base::NoDestructor<std::vector<std::string>> g_vetted_hosts(
```
just use `base::MakeFixedFlatSet` and `base::Contains` instead of linear search (see https://groups.google.com/a/chromium.org/g/chromium-dev/c/lVsznaNa_Bk)

petemill replied:
ok thanks, I just copied this from the existing API to be used by both APIs, but happy to make the improvement

petemill replied:
fixed, thanks again

iefremov commented on components/weekly_storage/daily_storage.h:
```diff
@@ -0,0 +1,55 @@
+// Allows to track a sum of some
+// values added from time to time via |AddDelta| over the last 24 hours.
+// Requires |pref_name| to be already registered.
+class DailyStorage {
```
why don't we reuse weekly storage by passing an interval as a parameter? could be a template parameter basically. Note that weekly storage has unit tests. I'm not sure copypasting would be a best choice

petemill replied:
We possibly _could_ combine WeeklyStorage and DailyStorage, but I actually didn't copy/paste - DailyStorage is quite a bit simpler thank WeeklyStorage I think. I'm going to ask that we look in to combining them as a follow-up for consideration, as I think adding DailyStorage capability to WeeklyStorage could be seen as adding even more complexity to WeeklyStorage. Would that be acceptable to you? I've added unit tests to DailyStorage, please review.

bridiver commented on components/brave_search/browser/brave_search_default_host.cc:
```diff
@@ -0,0 +1,141 @@
+    : host_(host),
+      template_url_service_(template_url_service),
+      prefs_(prefs),
+      weak_factory_(this) {}
```
WeakPtr doesn't appear to ever be used

petemill replied:
fixed

bridiver commented on browser/brave_content_browser_client.cc:
```diff
@@ -220,11 +225,34 @@ void BindBraveSearchHost(
+  } else {
+    // Dummy API which always returns false for private contexts.
+    mojo::MakeSelfOwnedReceiver(
+        std::make_unique<brave_search::BraveSearchDefaultHostPrivate>(
```
do we actually need a different class for this? Can't we just return early if these values are null? Or alternatively why pass in the values at all if they don't matter?

petemill replied:
Well I just thought it might be a clean way to keep the private profile implementation separate from the regular profile, since there is no code overlap. Even if we allow the setting in private profile in the future, we'd have to use something similar to `kUseAlternativeSearchEngineProvider`. We need to pass in values because it inherits from `brave_search::mojom::BraveSearchDefault`, but I suppose I can provide an empty constructor in the subclass.

bridiver commented on browser/brave_content_browser_client.cc:
```diff
@@ -220,11 +225,34 @@ void BindBraveSearchHost(
+    mojo::PendingReceiver<brave_search::mojom::BraveSearchDefault> receiver) {
+  auto* context = frame_host->GetBrowserContext();
+  auto* profile = Profile::FromBrowserContext(context);
+  auto* template_url_service =
```
it would be better to check this inside `IsRegularProfile` so it doesn't go unused

petemill replied:
fixed

bridiver commented on browser/brave_profile_prefs.cc:
```diff
@@ -339,6 +340,10 @@ void RegisterProfilePrefs(user_prefs::PrefRegistrySyncable* registry) {
 #endif
 
+  // Brave Search
+  registry->RegisterListPref(brave_search::prefs::kDailyAsked);
```
these should go in `BraveSearchHost::RegisterProfilePrefs` or similar and call that here instead

petemill replied:
fixed

bridiver commented on components/brave_search/browser/brave_search_default_host.cc:
```diff
@@ -0,0 +1,141 @@
+                                      const std::string& host) {
+  // Prefer built-in entries, then offer site-provided entries,
+  // but ignore extension-provided entries.
+  TemplateURL::TemplateURLVector urls = service->GetTemplateURLs();
```
@simonhong do we need any null checks here?

simonhong replied:
|urls| only includes valid providers.
Upstream also iterates urls like below from many places.
```
for (const auto& turl : template_urls_) {
    if (turl->xxx())
```
```

bridiver commented on components/brave_search/renderer/brave_search_renderer_utils.cc:
```diff
@@ -0,0 +1,36 @@
+namespace brave_search {
+
+bool IsAllowedHost(const GURL& url) {
+  if (!url.SchemeIs("https")) {
```
url_constants `kHttpsScheme`

petemill replied:
fixed

bridiver commented on components/brave_search/renderer/brave_search_render_frame_observer.cc:
```diff
@@ -0,0 +1,50 @@
+  GURL url =
+      url::Origin(render_frame()->GetWebFrame()->GetSecurityOrigin()).GetURL();
+
+  if (!url.is_valid() || !url.SchemeIsHTTPOrHTTPS() ||
```
these checks are redundant since they are done in IsAllowedHost and also we don't allow http

petemill replied:
fixed via https://github.com/brave/brave-core/compare/ce5bad2..0c92e3644999fdf0b18efa432d5a07778dc2d7b7

petemill commented:
Please re-review. Changes:

- Addressed feedback
- Added more unit tests
- Implemented a feature and feature parameters for daily and total limits

Changed code since last reviews: https://github.com/brave/brave-core/compare/ce5bad2..0c92e3644999fdf0b18efa432d5a07778dc2d7b7

petemill commented:
CI issues unrelated, build is ✅ on all platforms

ShivanKaul commented on components/brave_search/browser/brave_search_default_host.cc:
```diff
@@ -0,0 +1,162 @@
+  if (is_default) {
+    return false;
+  }
+  if (!template_url_service_->CanMakeDefault(provider)) {
```
Is this where the Tor/private mode check is carried out?

petemill replied:
@ShivanKaul no, for those profiles, `BraveSearchDefaultHost` won't be created. Instead `BraveSearchDefaultHostPrivate` will, which always returns `false`: https://github.com/brave/brave-core/pull/8819/files#diff-594631d3a40b383b831cd6c6c401ff278c4b259ec7f1b321808e868ccfb06f2bR252-R257

bridiver commented on components/brave_search/browser/brave_search_default_host.h:
```diff
@@ -0,0 +1,53 @@
+#include <memory>
+#include <string>
+
+#include "base/memory/weak_ptr.h"
```
this doesn't seem to be used?

bridiver commented on components/brave_search/browser/brave_search_default_host_private.cc:
```diff
@@ -0,0 +1,25 @@
+
+#include "brave/components/brave_search/browser/brave_search_default_host_private.h"
+
+#include <string>
```
I think most of these includes are not used?

bridiver commented on components/brave_search/browser/brave_search_default_host_private.h:
```diff
@@ -0,0 +1,38 @@
+#ifndef BRAVE_COMPONENTS_BRAVE_SEARCH_BROWSER_BRAVE_SEARCH_DEFAULT_HOST_PRIVATE_H_
+#define BRAVE_COMPONENTS_BRAVE_SEARCH_BROWSER_BRAVE_SEARCH_DEFAULT_HOST_PRIVATE_H_
+
+#include <memory>
```
same as above

mkarolin reviewed (approved):
`chromium_src` ++

petemill commented:
Android failure seems unrelated, in `create_dist`:
```
 FAILED: gen/android_webview/cancellation_signal_android_jar_jni_headers/CancellationSignal_jni.h 
```

Macos failure seems odd, in `build`:
```
17:15:07  In file included from ../../brave/chromium_src/content/browser/browser_main_loop.cc:8:
17:15:07  In file included from ../../brave/browser/brave_browser_main_loop.cc:8:
17:15:07  In file included from ../../brave/browser/brave_browser_main_parts.h:10:
17:15:07 
 ../../chrome/browser/chrome_browser_main.h:19:10: fatal error: 'chrome/common/buildflags.h' file not found
17:15:07  #include "chrome/common/buildflags.h"
17:15:07           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Will re-run macos.

bsclifton reviewed (approved):
Will be admin merging as we need this in Nightly 😄

SergeyZhukovsky commented on browser/brave_content_browser_client.cc:
```diff
@@ -236,11 +242,33 @@ void BindBraveSearchHost(
 }
+
+void BindBraveSearchDefaultHost(
+    content::RenderFrameHost* const frame_host,
```
we shouldn't pass `RenderFrameHost` here directly as the Bind is async it could be invalid at that moment and will crash. We should use `process_id` like here https://github.com/brave/brave-core/pull/8819/files#diff-594631d3a40b383b831cd6c6c401ff278c4b259ec7f1b321808e868ccfb06f2bR236

SergeyZhukovsky replied:
nvm, we are good here as rfh is not bound in `RegisterBrowserInterfaceBindersForFrame`, it comes here

SergeyZhukovsky commented on components/brave_search/renderer/brave_search_render_frame_observer.cc:
```diff
@@ -0,0 +1,49 @@
+    int32_t world_id)
+    : RenderFrameObserver(render_frame), world_id_(world_id) {
+  native_javascript_handle_.reset(
+      new BraveSearchDefaultJSHandler(render_frame));
```
we should create that object in `DidCreateScriptContext` as `render_frame` could be different or invalid while there. And if we bind remote to it, it will fail on mojo calls. Check a fix that is done here https://github.com/brave/brave-core/pull/9127/files#diff-bd0508db5780666e635128717e778f2a51486f055feb9f236d03aefdb1dc122cR35
